addons/mail/models/mail_notification.py

```diff
@@ -59,6 +59,11 @@ def init(self):
             CREATE INDEX IF NOT EXISTS mail_notification_res_partner_id_is_read_notification_status_mail_message_id
                                     ON mail_notification (res_partner_id, is_read, notification_status, mail_message_id)
         """)
+        self.env.cr.execute(
+            """CREATE UNIQUE INDEX IF NOT EXISTS unique_mail_message_id_res_partner_id_if_set
```

Comment:
Commit:

> [IMP] mail: add constraint at db level for partner / message notification

A small test case for this would maybe be nice? Not sure if easy to do.